### PR TITLE
Fixes #1672 + tests. If objects are neighbors, distances don't include touching

### LINKED ISSUE #1672: Docker build (tag:latest) has CP failures
Hello,

I'm using the cellprofiler/cellprofiler:latest docker version (updated a few hours ago), and get errors like this where I had none before using `:master`:

```
Did not find ij1-patcher.jar
Version: 2015-10-29T13:53:19 d998c87 / 20151029135319
Failed to stop Ilastik
Pipeline saved with CellProfiler version 20140723174500
ilastik import: failed to import the ilastik. Please follow the instructions on 
    "http://www.ilastik.org" to install ilastik
Traceback (most recent call last):
  File "/usr/local/src/CellProfiler/cellprofiler/modules/classifypixels.py", line 82, in <module>
    from ilastik.core.dataMgr import DataMgr, DataItemImage
ImportError: No module named ilastik.core.dataMgr
ilastik import: failed to import the ilastik. Please follow the instructions on 
                          "http://www.ilastik.org" to install ilastik
Traceback (most recent call last):
  File "/usr/local/src/CellProfiler/cellprofiler/modules/ilastik_pixel_classification.py", line 81, in <module>
    import ilastik_main
ImportError: No module named ilastik_main
Times reported are CPU times for each module, not wall-clock time
Thu Oct 29 17:09:57 2015: Image # 15177, module Images # 1: 0.00 sec
Thu Oct 29 17:09:57 2015: Image # 15177, module Metadata # 2: 0.00 sec
Warning:  org.apache.xerces.parsers.SAXParser: Feature 'http://javax.xml.XMLConstants/feature/secure-processing' is not recognized.
Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://javax.xml.XMLConstants/property/accessExternalDTD' is not recognized.
Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit' is not recognized.
CP-JAVA 17:09:59.944 [Thread-0] ERROR loci.formats.ClassList - "loci.formats.in.SlideBook6Reader" is not valid.
CP-JAVA 17:10:00.684 [Thread-0] ERROR loci.formats.ClassList - "loci.formats.in.SlideBook6Reader" is not valid.
Exception in thread "Thread-0" java.lang.NullPointerException
    at loci.formats.in.ZeissLMSReader.isThisType(ZeissLMSReader.java:76)
Caught exception in class loci.formats.in.ZeissLMSReader.isThisType
CP-JAVA 17:10:03.010 [Thread-0] ERROR loci.formats.ClassList - "loci.formats.in.SlideBook6Reader" is not valid.
Exception in thread "Thread-0" java.lang.NullPointerException
    at loci.formats.in.ZeissLMSReader.isThisType(ZeissLMSReader.java:76)
Caught exception in class loci.formats.in.ZeissLMSReader.isThisType
CP-JAVA 17:10:03.869 [Thread-0] ERROR loci.formats.ClassList - "loci.formats.in.SlideBook6Reader" is not valid.
Exception in thread "Thread-0" java.lang.NullPointerException
    at loci.formats.in.ZeissLMSReader.isThisType(ZeissLMSReader.java:76)
Caught exception in class loci.formats.in.ZeissLMSReader.isThisType
CP-JAVA 17:10:04.446 [Thread-0] ERROR loci.formats.ClassList - "loci.formats.in.SlideBook6Reader" is not valid.
Exception in thread "Thread-0" java.lang.NullPointerException
    at loci.formats.in.ZeissLMSReader.isThisType(ZeissLMSReader.java:76)
Caught exception in class loci.formats.in.ZeissLMSReader.isThisType
Thu Oct 29 17:09:58 2015: Image # 15177, module NamesAndTypes # 3: 3.99 sec
Thu Oct 29 17:10:05 2015: Image # 15177, module Groups # 4: 0.00 sec
Thu Oct 29 17:10:05 2015: Image # 15177, module IdentifyPrimaryObjects # 5: 6.03 sec
Thu Oct 29 17:10:15 2015: Image # 15177, module IdentifySecondaryObjects # 6: 12.53 sec
Thu Oct 29 17:10:38 2015: Image # 15177, module IdentifyTertiaryObjects # 7: 1.13 sec
Thu Oct 29 17:10:39 2015: Image # 15177, module MeasureObjectIntensity # 8: 28.77 sec
Error detected during run of module MeasureObjectNeighbors
Traceback (most recent call last):
  File "/usr/local/src/CellProfiler/cellprofiler/pipeline.py", line 1830, in run_with_yield
    self.run_module(module, workspace)
  File "/usr/local/src/CellProfiler/cellprofiler/pipeline.py", line 2078, in run_module
    module.run(workspace)
  File "/usr/local/src/CellProfiler/cellprofiler/modules/measureobjectneighbors.py", line 440, in run
    distance_matrix[:, ~neighbor_has_pixels] = np.inf
IndexError: index 442 is out of bounds for axis 1 with size 435
Thu Oct 29 17:11:09 2015: Image # 15177, module MeasureObjectNeighbors # 9: 1.77 sec
Failed to stop Ilastik
```

If `:latest` is actually a bleeding-edge build, then I probably shouldn't be using it. However I'm not sure if this tag is supposed to work and needs fixing, or if I should just be using `:master` instead.

Thanks!

### PR DESCRIPTION
**Pull request by LeeKamentsky**
I introduced a bug in 9a2b6a6dc9981ddc0b5d566109067238d12fa78b that shows up when you calculate distances to the nearest neighbor. The distance matrix is computed excluding the objects touching the image border, but the neighbor_has_pixels array was calculated using objects touching the border as well (and these have larger object numbers than the ones that don't). The array size was wrong and the code failed as indicated.

@0x00B1 - this pull should get a little priority since the reporter is an ASTP client.

### CODE REVIEW

**Comment by 0x00B1** [comment]
I don’t see any issues, but what’s going on with Travis?

```sh
/home/travis/virtualenv/python2.7_with_system_site_packages/local/lib/python2.7/site-packages/numpy/core/fromnumeric.py:2641: VisibleDeprecationWarning: `rank` is deprecated; use the `ndim` attribute or function instead. To find the rank of a matrix see `numpy.linalg.matrix_rank`.
```

File: cellprofiler/modules/tests/test_measureobjectneighbors.py [hunk 1]
```diff
@@ -806,4 +806,27 @@ def test_08_03_object_is_missing(self):
         ftr = module.get_measurement_name(M.M_FIRST_CLOSEST_OBJECT_NUMBER)
         values = m[OBJECTS_NAME, ftr]
         self.assertEqual(values[1], 1)
-
```
Comment (0x00B1) [inline cellprofiler/modules/tests/test_measureobjectneighbors.py]: Typo. :stuck_out_tongue_winking_eye: You can remove this from the commit.

**Comment by 0x00B1** [comment]
I suspect a new NumPy package was released. I guess we should fix the warning.

**Comment by LeeKamentsky** [comment]
h5py asks for numpy > 1.6.1 and somehow pip decides it should uninstall numpy 1.9.1 and install numpy 1.10.1. But prior to that, binaries had been installed that depend on numpy 1.9.1 and may have been built using the numpy.h file from 1.9.1. If the structures changed, then those binaries will be incompatible. I think that's where some of the error messages are coming from, others are because it's using 1.10.1.

See this line: https://travis-ci.org/CellProfiler/CellProfiler/builds/88205633#L311

I am looking for numpy requirements in that logfile.

File: cellprofiler/modules/tests/test_measureobjectneighbors.py (hunk 1 above)
Comment (LeeKamentsky) [inline cellprofiler/modules/tests/test_measureobjectneighbors.py]: How do I correct that? It sure is an ugly diff.

**Comment by LeeKamentsky** [comment]
I've squashed things so that there are two commits - one to fix the original problem and one to get travis to build correctly. I did that so that you could cherry-pick just the changes that fix MeasureNeighbors.

**Comment by 0x00B1** [comment]
Thanks!

However, I don’t think cherry picking is necessary since the pull request documents your reasoning for the additional commit (i.e. your Travis CI fixes).

**Comment by purohit** [comment]
Thanks for the fix @LeeKamentsky! Any idea when Docker is going to build a version with this fix in it? Says the last build was 5 days ago.

**Comment by LeeKamentsky** [comment]
I just triggered a build manually and it looks like it's there now. Let me know if it's not.

**Pull request merged**